fuel/datasets/hdf5.py

```diff
@@ -38,24 +45,27 @@ def __init__(self, sources, start, stop, path, data_node='Data',
         self.provides_sources = sources
         self.path = path
         self.data_node = data_node
+        self.open_file(path)
         self.start = start
         self.stop = stop
-        self.num_examples = self.stop - self.start
         self.nodes = None
-        self.open_file(self.path)
-        super(Hdf5Dataset, self).__init__(self.provides_sources)
+        self.open_file(path)
+        super(PytablesDataset, self).__init__(self.provides_sources)
 
     def open_file(self, path):
         self.h5file = tables.open_file(path, mode="r")
         node = self.h5file.getNode('/', self.data_node)
 
         self.nodes = [getattr(node, source) for source in self.sources_in_file]
+        if self.stop is None:
+            self.stop = self.nodes[0].nrows
+        self.num_examples = self.stop - self.start
 
     def load(self):
         self.open_file(self.path)
 
-    def close(self):
-        self.h5file.close()
+    def close_file(self):
+        self.h5file.close_file()
```

Comment:
We should also be deleting `self._h5file` and `self.nodes`, so that they're properly reloaded if the dataset tries to access these attributes again in its lifetime.